src/main/drivers/max7456.c

```diff
@@ -155,24 +156,48 @@
 // On shared SPI buss we want to change clock for OSD chip and restore for other devices.
 
 #ifdef MAX7456_SPI_CLK
-    #define ENABLE_MAX7456        {spiSetDivisor(MAX7456_SPI_INSTANCE, MAX7456_SPI_CLK);IOLo(max7456CsPin);}
+    #define ENABLE_MAX7456()    do { \
+                                    spiSetDivisor(MAX7456_SPI_INSTANCE, MAX7456_SPI_CLK); \
+                                    IOLo(max7456CsPin); \
+                                } while(0)
 #else
-    #define ENABLE_MAX7456        IOLo(max7456CsPin)
+    #define ENABLE_MAX7456()    do { \
+                                    IOLo(max7456CsPin); \
+                                } while(0)
 #endif
 
 #ifdef MAX7456_RESTORE_CLK
-    #define DISABLE_MAX7456       {IOHi(max7456CsPin);spiSetDivisor(MAX7456_SPI_INSTANCE, MAX7456_RESTORE_CLK);}
+    #define DISABLE_MAX7456()   do { \
+                                    IOHi(max7456CsPin); \
+                                    spiSetDivisor(MAX7456_SPI_INSTANCE, MAX7456_RESTORE_CLK); \
+                                } while(0)
 #else
-    #define DISABLE_MAX7456       IOHi(max7456CsPin)
+    #define DISABLE_MAX7456()   do { \
+                                    IOHi(max7456CsPin); \
+                                } while(0)
 #endif
 
 uint16_t maxScreenSize = VIDEO_BUFFER_CHARS_PAL;
 
-// we write everything in screenBuffer and then comapre
-// screenBuffer with shadowBuffer to upgrade only changed chars
-// this solution is faster then redraw all screen
-static uint8_t screenBuffer[VIDEO_BUFFER_CHARS_PAL+40]; //for faster writes we use memcpy so we need some space to don't overwrite buffer
-static uint8_t shadowBuffer[VIDEO_BUFFER_CHARS_PAL];
+// we write everything in screenBuffer and set a dirty bit
+// in screenIsDirty to upgrade only changed chars this solution
+// is faster than redrawing the whole screen on each frame
+static uint8_t screenBuffer[VIDEO_BUFFER_CHARS_PAL];
+static uint8_t screenIsDirty[VIDEO_BUFFER_CHARS_PAL/8] = {0,};
+
+STATIC_ASSERT(VIDEO_BUFFER_CHARS_PAL % 4 == 0, VIDEO_BUFFER_CHARS_PAL_not_multiple_of_4);
+// Needs to be multiple of 32 because we're using 1 bit per
+// character in screenIsDirty and then setting it via aliasing
+// to uint32_t.
+STATIC_ASSERT(VIDEO_BUFFER_CHARS_PAL % 32 == 0, VIDEO_BUFFER_CHARS_PAL_not_multiple_of_32);
+
+#define BYTE_ARRAY_SET_BY_WORDS(arr, x) \
```

Comment:
`memset` of newlib-nano is smart enough to use 32-bit access for fastest possible operation. The only downside of using `memset` is a function call overhead.

Reply:
I hadn't bothered checking the source of `memset`. I assumed the code setting the buffer by aliasing to a 32bit ptr was there for some good reason, but now that you mention it I think it might be a leftover from the MAX7456 handling in the Arduino-based projects. Using `memset` saves 32 bits of flash, so let's change it.

Reply:
@fiam Since both `screenBuffer` and `screenIsDirty` are `uint8_t` arrays and it would be good to add an `__attribute__ ((aligned (4)))` to their definition to make sure `memset` won't have to deal with mis-aligned bytes.